components/brave_wallet/browser/eth_allowance_manager.cc

```diff
@@ -0,0 +1,329 @@
+      json_rpc_service_(json_rpc_service),
+      keyring_service_(keyring_service),
+      prefs_(prefs),
+      weak_ptr_factory_(this) {}
+
+EthAllowanceManager::~EthAllowanceManager() = default;
+
+void EthAllowanceManager::DiscoverEthAllowancesOnAllSupportedChains() {
+  if (is_eth_allowance_discovering_running_) {
+    return;
+  }
+  is_eth_allowance_discovering_running_ = true;
+
+  const auto keyring_info = keyring_service_->GetKeyringInfoSync(
+      brave_wallet::mojom::kDefaultKeyringId);
+  std::vector<std::string> account_addresses;
+  for (const auto& account_info : keyring_info->account_infos) {
+    account_addresses.push_back(account_info->address);
+  }
+
+  if (account_addresses.empty()) {
+    return;
+  }
+
+  const auto token_list_map =
+      BlockchainRegistry::GetInstance()->GetEthTokenListMap(
+          GetChainIdsForAlowanceDiscovering());
+
+  base::flat_map<std::string, base::Value::List> chain_id_to_contract_addresses;
+
+  for (const auto& [chain_id, token_list] : token_list_map) {
+    for (const auto& token : token_list) {
+      chain_id_to_contract_addresses[chain_id].Append(token->contract_address);
+    }
+  }
+
+  if (chain_id_to_contract_addresses.empty()) {
+    return;
+  }
+  const auto& allowance_cashe_dict =
+      prefs_->GetDict(kBraveWalletEthAllowancesCache);
+
+  // Use a barrier callback to wait for all EthGetLogs calls to
+  // complete (one for each chain_id).
+  const auto barrier_callback = base::BarrierCallback<base::flat_map<
+      std::string,
+      std::tuple<uint256_t, std::vector<mojom::AllowanceInfoPtr>>>>(
+      account_addresses.size() * chain_id_to_contract_addresses.size(),
+      base::BindOnce(&EthAllowanceManager::MergeEthAllowances,
+                     weak_ptr_factory_.GetWeakPtr()));
+
+  const auto approval_topic_hash =
+      KeccakHash("Approval(address,address,uint256)");
+  for (const auto& account_address : account_addresses) {
+    for (auto& [chain_id, contract_addresses] :
+         chain_id_to_contract_addresses) {
+      const auto* chain_cashed_data = allowance_cashe_dict.FindDict(chain_id);
+
+      const auto* last_block_number_ptr =
+          chain_cashed_data != nullptr
+              ? chain_cashed_data->FindString("last_block_number")
+              : nullptr;
+      const std::string last_block_number{
+          last_block_number_ptr != nullptr ? *last_block_number_ptr : ""};
+
+      std::string hex_account_address;
+      if (!PadHexEncodedParameter(account_address, &hex_account_address)) {
+        continue;
+      }
+      auto internal_callback = base::BindOnce(
+          &EthAllowanceManager::OnGetAllowances, weak_ptr_factory_.GetWeakPtr(),
+          barrier_callback, chain_id, hex_account_address);
+
+      base::Value::List topics;
+      topics.Append(approval_topic_hash);
+      topics.Append(std::move(hex_account_address));
+
+      base::Value::Dict filter_options;
+      filter_options.Set("address", std::move(contract_addresses));
```

Comment:
If you move this, wouldn't next account_address loop don't have the right value of contract_addresses to use?
Need a test case to ensure things are working correctly for multiple accounts if we don't have it yet.